compiler/builddeps.nim

```diff
@@ -0,0 +1,48 @@
+#[
+## TODO
+* this will show `CC: dragonbox`:
+`nim r -f --hint:cc --filenames:abs --processing:filenames --hint:conf nonexistant`
+we could refine the logic to delay compilation until cgen phase instead.
+(see also `tests/osproc/treadlines.nim`)
+* allow some form of reporting so that caller can tell whether a dependency
+  doesn't exist, or was already built, or builds with error, or builds successfully, etc.
+]#
+
+import std/[osproc, os, strutils]
+import msgs, options, ast, lineinfos, extccomp, pathutils
+
+const prefix = "__" # prevent clashing with user files
+
+proc addDependency*(conf: ConfigRef, name: string, info: TLineInfo) =
+  case name
+  of "dragonbox":
+    if name notin conf.dependencies:
+      conf.dependencies.add name
+      # xxx we could also build this under $nimb/build/
+      let dir = conf.getNimcacheDir().string
+      createDir dir
+      let objFile = dir / ("$1nimdragonbox.o" % prefix)
+      if optForceFullMake in conf.globalOptions or not objFile.fileExists:
+        # xxx
+        # let cppExe = getCompilerExe(c.config; compiler: TSystemCC; cfile: AbsoluteFile): string =
```

Comment:
I'm currently hardcoding the C++ compiler; followup PR will improve this to use `extccomp.getCompilerExe` instead so it honors user configs (deferring this to future PR as it requires a few other changes to allow this and this PR is already quite big)

Reply:
We should c2nim the code, as I said.

Reply:
as I wrote above, I've tried that already; c2nim can't handle dragonbox.cc, it lacks a lot of functionality to support C++ files like dragonbox.cc and chokes on many different parts of the code.
Using dragonbox.cc directly is much simpler than making c2nim support C++ properly, which could easily represent a multi-year effort.